### PR TITLE
refactor: Cleanup thread ctor calls

### PR DESCRIPTION
**Pull request by hebasto**
This PR does not change behavior.
Its goal is to improve readability and maintainability of the code.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Concept ACK to enforce the void return type on threads

**Comment by Empact** [comment]
Concept ACK on the first two commits, ~0 on https://github.com/bitcoin/bitcoin/pull/19064/commits/2a816bd79093eac7d0bbae3106fdd7fddbc99cf2 as I don't see the cited improvements justifying the change.

File: src/util/system.cpp [hunk 1]
```diff
@@ -1132,6 +1132,25 @@ int GetNumCores()
     return std::thread::hardware_concurrency();
 }
 
+void TraceThread(const char* thread_name, std::function<void()> thread_func)
```
Comment (MarcoFalke) [inline src/util/system.cpp:1135]: As you move this anyway, what about moving threadnames.h to thread.h and adding this to that file? Advantage would be that stuff like `src/index/base.cpp` that doesn't need the whole system.h can just include the slim thread.h

**Comment by hebasto** [comment]
Updated 2a816bd79093eac7d0bbae3106fdd7fddbc99cf2 -> bcd04708a6fe32cf7fba2969ec378c34e6950278 ([pr19064.01](https://github.com/hebasto/bitcoin/commits/pr19064.01) -> [pr19064.03](https://github.com/hebasto/bitcoin/commits/pr19064.03), [diff](https://github.com/hebasto/bitcoin/compare/pr19064.01..pr19064.03)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/19064#discussion_r431512796):
> As you move this anyway, what about moving threadnames.h to thread.h and adding this to that file? Advantage would be that stuff like `src/index/base.cpp` that doesn't need the whole system.h can just include the slim thread.h

File: src/util/thread.h [hunk 2]
```diff
@@ -21,6 +22,11 @@ void ThreadSetInternalName(std::string&&);
 //! logging.
 const std::string& ThreadGetInternalName();
 
+/**
+ * A wrapper for do-something-once thread functions.
```
Comment (MarcoFalke) [inline src/util/thread.h:26]: Based on the circular dependency and what this function does (Run a thread with log statements), maybe this is better suited in `logging.h`? No strong opinion, so feel free to not move this at all.

Comment (hebasto) [reply src/util/thread.h:26]: Maybe keep `threadnames.{h,cpp}` _and_ move `TraceThread()` to new `thread.{h,cpp}`?

Comment (MarcoFalke) [reply src/util/thread.h:26]: Sure if that works. Cutting down the util/system bloat can't hurt.

**Comment by hebasto** [comment]
Updated bcd04708a6fe32cf7fba2969ec378c34e6950278 -> 90aa1f7f503818a67d18ec20401707d5e4ca658d ([pr19064.03](https://github.com/hebasto/bitcoin/commits/pr19064.03) -> [pr19064.04](https://github.com/hebasto/bitcoin/commits/pr19064.04), [diff](https://github.com/hebasto/bitcoin/compare/pr19064.03..pr19064.04)):

- fixed circular dependencies

File: src/util/thread.h (hunk 2 above)
Comment (hebasto) [reply src/util/thread.h:12]: [Updated](https://github.com/bitcoin/bitcoin/pull/19064#issuecomment-636361588).

**Comment by hebasto** [comment]
Rebased 90aa1f7f503818a67d18ec20401707d5e4ca658d -> dd00f7c2c3d3af64f8a09d6b36d80f80fa96ff48 ([pr19064.04](https://github.com/hebasto/bitcoin/commits/pr19064.04) -> [pr19064.05](https://github.com/hebasto/bitcoin/commits/pr19064.05)) due to the conflict with #19180.

**Comment by hebasto** [comment]
Updated dd00f7c2c3d3af64f8a09d6b36d80f80fa96ff48 -> 73370d039ec42a71171569e9b3f012e936a4ebe3 ([pr19064.05](https://github.com/hebasto/bitcoin/commits/pr19064.05) -> [pr19064.06](https://github.com/hebasto/bitcoin/commits/pr19064.06), [diff](https://github.com/hebasto/bitcoin/compare/pr19064.05..pr19064.06)):

- fixed Windows builds

**Comment by hebasto** [comment]
Rebased 73370d039ec42a71171569e9b3f012e936a4ebe3 -> ea12b8f04dc869f740c41fdc28d9d9fb6486f20c ([pr19064.06](https://github.com/hebasto/bitcoin/commits/pr19064.06) -> [pr19064.07](https://github.com/hebasto/bitcoin/commits/pr19064.07)) due to the conflicts with #19028 and #19197.

**Comment by hebasto** [comment]
Rebased ea12b8f04dc869f740c41fdc28d9d9fb6486f20c -> 427e396b16564d9c735238a2b2d24472bfcd55c3 ([pr19064.07](https://github.com/hebasto/bitcoin/commits/pr19064.07) -> [pr19064.08](https://github.com/hebasto/bitcoin/commits/pr19064.08)) due to the conflict with #15382.

**Comment by hebasto** [comment]
Rebased 427e396b16564d9c735238a2b2d24472bfcd55c3 -> 338488f056169898905bf03fb3df34e6877833ac ([pr19064.08](https://github.com/hebasto/bitcoin/commits/pr19064.08) -> [pr19064.09](https://github.com/hebasto/bitcoin/commits/pr19064.09)) due to the conflict with #19316.

File: src/txmempool.cpp [hunk 3]
```diff
@@ -19,6 +19,8 @@
 #include <util/time.h>
 #include <validationinterface.h>
 
+#include <math.h>
```
Comment (practicalswift) [inline src/txmempool.cpp:22]:
```suggestion
#include <cmath>
```

**Review by practicalswift — COMMENTED** [review]
Concept ACK

**Comment by hebasto** [comment]
Updated 338488f056169898905bf03fb3df34e6877833ac -> 4335a15ac6aca65430471ead52d389813f4260e7 ([pr19064.09](https://github.com/hebasto/bitcoin/commits/pr19064.09) -> [pr19064.10](https://github.com/hebasto/bitcoin/commits/pr19064.10), [diff](https://github.com/hebasto/bitcoin/compare/pr19064.09..pr19064.10)):

- addressed @practicalswift's [comment](https://github.com/bitcoin/bitcoin/pull/19064#discussion_r471075241)

File: src/txmempool.cpp (hunk 3 above)
Comment (hebasto) [reply src/txmempool.cpp]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19064#issuecomment-674491163).

**Comment by hebasto** [comment]
Rebased 4335a15ac6aca65430471ead52d389813f4260e7 -> 806aedebc2b0627e681480be39c2d93af87e0863 ([pr19064.10](https://github.com/hebasto/bitcoin/commits/pr19064.10) -> [pr19064.11](https://github.com/hebasto/bitcoin/commits/pr19064.11)) due to the conflict with #19779.

**Comment by hebasto** [comment]
Rebased 806aedebc2b0627e681480be39c2d93af87e0863 -> 8cb5f901dfeb8de5427532e6646c3a5a131fed14 ([pr19064.11](https://github.com/hebasto/bitcoin/commits/pr19064.11) -> [pr19064.12](https://github.com/hebasto/bitcoin/commits/pr19064.12)) due to the conflict with #19991.

**Comment by hebasto** [comment]
Rebased 8cb5f901dfeb8de5427532e6646c3a5a131fed14 -> 7e7a999e9e4bdef9726037afa4ec0325a355528e ([pr19064.12](https://github.com/hebasto/bitcoin/commits/pr19064.12) -> [pr19064.13](https://github.com/hebasto/bitcoin/commits/pr19064.13)).

**Comment by hebasto** [comment]
Rebased 7e7a999e9e4bdef9726037afa4ec0325a355528e -> 25c8e21b0f316ce7804e10de813c4bcd81d3c2f3 ([pr19064.13](https://github.com/hebasto/bitcoin/commits/pr19064.13) -> [pr19064.14](https://github.com/hebasto/bitcoin/commits/pr19064.14)).

**Comment by hebasto** [comment]
Rebased 25c8e21b0f316ce7804e10de813c4bcd81d3c2f3 -> 358a0447f59608af7bd51fe200e27baefb78159e ([pr19064.14](https://github.com/hebasto/bitcoin/commits/pr19064.14) -> [pr19064.15](https://github.com/hebasto/bitcoin/commits/pr19064.15)) due to the conflict with #20323.

**Comment by hebasto** [comment]
Updated 358a0447f59608af7bd51fe200e27baefb78159e -> 519838dbb91410544743b69c3d386fff9e4eb400 ([pr19064.15](https://github.com/hebasto/bitcoin/commits/pr19064.15) -> [pr19064.16](https://github.com/hebasto/bitcoin/commits/pr19064.16)):

- rebased due to the conflict with #18077
- addressed @Empact's [comment](https://github.com/bitcoin/bitcoin/pull/19064#issuecomment-634329285) and dropped the last controversial commit

**Review by jonatack — COMMENTED** [review]
Concept ACK. Rebased to current master 15a9df070 and this debug builds cleanly. Now that we have moved on from C++ 11, can this pull also use lambdas in the place of `std::bind` and `std::function`?

**Comment by hebasto** [comment]
> Concept ACK. Rebased to current master [15a9df0](https://github.com/bitcoin/bitcoin/commit/15a9df070615e7e8f29c17a92b889f19218f25ac) and this debug builds cleanly. Now that we have moved on from C++ 11, can this pull also use lambdas in the place of `std::bind` and `std::function`?

The dedicated commit was removed from this PR, see https://github.com/bitcoin/bitcoin/pull/19064#issuecomment-634329285

**Comment by jonatack** [comment]
It seems worthwhile to do. https://youtu.be/ZlHi8txU4aQ

**Comment by jonatack** [comment]
Also https://youtu.be/zt7ThwVfap0?t=1721 (I'm still going through it, though)

Screenshot from 2021-01-27 18-21-18

Screenshot from 2021-01-27 18-20-48

**Comment by hebasto** [comment]
Updated 519838dbb91410544743b69c3d386fff9e4eb400 -> 33f9134f61d99903939c3048340927d051fd60f9 ([pr19064.16](https://github.com/hebasto/bitcoin/commits/pr19064.16) -> [pr19064.17](https://github.com/hebasto/bitcoin/commits/pr19064.17)):

- rebased due to the numerous conflicts
- addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/19064#pullrequestreview-577475338)

File: src/validation.cpp [hunk 4]
```diff
@@ -46,6 +46,7 @@
 #include <util/rbf.h>
 #include <util/strencodings.h>
 #include <util/system.h>
+#include <util/threadnames.h>
```
Comment (kiminuo) [inline src/validation.cpp:49]: Why is this added? I can't find any `util::*` call.

**Comment by hebasto** [comment]
Updated 33f9134f61d99903939c3048340927d051fd60f9 -> 2a3ad116c8cf00ac44c9ad6c8585edf37054761d ([pr19064.17](https://github.com/hebasto/bitcoin/commits/pr19064.17) -> [pr19064.18](https://github.com/hebasto/bitcoin/commits/pr19064.18), [diff](https://github.com/hebasto/bitcoin/compare/pr19064.17..pr19064.18)):

- addressed @kiminuo's [comment](https://github.com/bitcoin/bitcoin/pull/19064#discussion_r619778521)

File: src/validation.cpp (hunk 4 above)
Comment (hebasto) [reply src/validation.cpp]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19064#issuecomment-826291903).

File: src/util/system.cpp (hunk 1 above)
Comment (hebasto) [reply src/util/system.cpp]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19064#issuecomment-635240245).

**Review by jonatack — COMMENTED** [review]
ACK 2a3ad116c8cf00ac44c9ad6c8585edf37054761d

Recommend reviewers view the diff squashed to one commit as the second and third commit rewrite parts of the first one.

File: src/init.cpp [hunk 5]
```diff
@@ -1266,7 +1267,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     node.scheduler = std::make_unique<CScheduler>();
 
     // Start the lightweight task scheduler thread
-    node.scheduler->m_service_thread = std::thread([&] { TraceThread("scheduler", [&] { node.scheduler->serviceQueue(); }); });
+    node.scheduler->m_service_thread = std::thread([&] { util::TraceThread("scheduler", [&] { node.scheduler->serviceQueue(); }); });
```
Comment (jnewbery) [inline src/init.cpp:1270]: Perhaps:

```suggestion
    node.scheduler->m_service_thread = std::thread(util::TraceThread, "scheduler", [&] { node.scheduler->serviceQueue(); });
```

To make this more readable and consistent with the other calls to `std::thread` ctor.

File: src/test/util/setup_common.cpp [hunk 6]
```diff
@@ -132,7 +134,7 @@ ChainTestingSetup::ChainTestingSetup(const std::string& chainName, const std::ve
     // We have to run a scheduler thread to prevent ActivateBestChain
     // from blocking due to queue overrun.
     m_node.scheduler = std::make_unique<CScheduler>();
-    m_node.scheduler->m_service_thread = std::thread([&] { TraceThread("scheduler", [&] { m_node.scheduler->serviceQueue(); }); });
+    m_node.scheduler->m_service_thread = std::thread([&] { util::TraceThread("scheduler", [&] { m_node.scheduler->serviceQueue(); }); });
```
Comment (jnewbery) [inline src/test/util/setup_common.cpp:137]: Perhaps:

```suggestion
    m_node.scheduler->m_service_thread = std::thread(&util::TraceThread("scheduler", [&] { m_node.scheduler->serviceQueue(); });
```

To make this more readable and consistent with the other calls to `std::thread` ctor.

**Review by jnewbery — COMMENTED** [review]
Looks good. A few suggestions inline.

Perhaps `thread.h` and `threadnames.h` should be merged?

File: src/txmempool.cpp [hunk 7]
```diff
@@ -18,6 +18,7 @@
 #include <validation.h>
 #include <validationinterface.h>
 
+#include <cmath>
```
[the remark this answers is not in the document]
Comment (jonatack) [reply src/txmempool.cpp:21]: I asked myself this as well--it's a missing header for `llround` used in the file.

[the remark this answers is not in the document]
Comment (hebasto) [reply src/txmempool.cpp:21]: In a previous iteration of this PR this header was required to build successfully for the reason @jonatack pointed at.

Currently, this change looks unrelated. Should I drop it?

**Comment by hebasto** [comment]
Updated 2a3ad116c8cf00ac44c9ad6c8585edf37054761d -> 1edb237b69fab42e1a8c77c605a7f94c653dd1d8 ([pr19064.18](https://github.com/hebasto/bitcoin/commits/pr19064.18) -> [pr19064.19](https://github.com/hebasto/bitcoin/commits/pr19064.19), [diff](https://github.com/hebasto/bitcoin/compare/pr19064.18..pr19064.19)):

- addressed all of @jnewbery's comments

File: src/txmempool.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (hebasto) [reply src/txmempool.cpp]: [Updated](https://github.com/bitcoin/bitcoin/pull/19064#issuecomment-829332687).

**Comment by hebasto** [comment]
Updated 2a3ad116c8cf00ac44c9ad6c8585edf37054761d -> 792be53d3e9e366b9f6aeee7a1eeb912fa28062e ([pr19064.18](https://github.com/hebasto/bitcoin/commits/pr19064.18) -> [pr19064.19](https://github.com/hebasto/bitcoin/commits/pr19064.19), [diff](https://github.com/hebasto/bitcoin/compare/pr19064.18..pr19064.19)):

- addressed @jnewbery's comments

@jnewbery 
> Perhaps `thread.h` and `threadnames.h` should be merged?

It will introduce a new circular dependency "logging -> util/threadnames -> logging".

File: src/net.cpp [hunk 8]
```diff
@@ -2546,15 +2547,14 @@ bool CConnman::Start(CScheduler& scheduler, const Options& connOptions)
         return false;
     }
     if (connOptions.m_use_addrman_outgoing || !connOptions.m_specified_outgoing.empty())
-        threadOpenConnections = std::thread(&TraceThread<std::function<void()> >, "opencon", std::function<void()>(std::bind(&CConnman::ThreadOpenConnections, this, connOptions.m_specified_outgoing)));
+        threadOpenConnections = std::thread(&util::TraceThread, "opencon", [this, connOptions] { ThreadOpenConnections(connOptions.m_specified_outgoing); });
```
[the remark this answers is not in the document]
Comment (hebasto) [reply src/net.cpp]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19064#issuecomment-829351149).

File: src/init.cpp (hunk 5 above)
Comment (hebasto) [reply src/init.cpp]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19064#issuecomment-829351149).

File: src/test/util/setup_common.cpp (hunk 6 above)
Comment (hebasto) [reply src/test/util/setup_common.cpp]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19064#issuecomment-829351149).

**Comment by jnewbery** [comment]
utACK 792be53d3e9e366b9f6aeee7a1eeb912fa28062e

**Comment by jonatack** [comment]
Changes are even nicer now.

tACK 792be53d3e9e366b9f6aeee7a1eeb912fa28062e

File: src/init.cpp [hunk 9]
```diff
@@ -1266,7 +1267,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     node.scheduler = std::make_unique<CScheduler>();
 
     // Start the lightweight task scheduler thread
-    node.scheduler->m_service_thread = std::thread([&] { TraceThread("scheduler", [&] { node.scheduler->serviceQueue(); }); });
+    node.scheduler->m_service_thread = std::thread(util::TraceThread, "scheduler", [&] { node.scheduler->serviceQueue(); });
```
Comment (MarcoFalke) [inline src/init.cpp:1270]: Is there a reason or even a difference between sometimes using `&util::TraceThread` and other times `util::TraceThread`?

**Comment by MarcoFalke** [comment]
cr ACK 792be53d3e9e366b9f6aeee7a1eeb912fa28062e

File: src/init.cpp (hunk 9 above)
Comment (hebasto) [reply src/init.cpp:1270]: No, there isn't. See https://en.cppreference.com/w/cpp/language/pointer#Pointers_to_functions

It seems I overlooked that inconsistency.

**Pull request merged**